project/02-Modularity/Handout.md

```diff
@@ -0,0 +1,160 @@
+
+> :information_source: You may notice while writing the implementation that the interface misses something: a method parameter, a list of thrown exceptions... This is not that big of a deal: sometimes, when writing the actual code, you come up with something you didn't think you'd need. Updating the interface at this time to provide what you need for that is not an issue. **However, please go through the same reasoning every time you update the interface as when you wrote it: is it specific to my implementation? Does it give too much details about the implementation? Do I really need it? ...**. 
+
+> :warning: In particular in today's project, if you realize you need a `Context` (or similar Android system classes) to do something, you should not add it as a method parameter in the interface, as it makes it tightly coupled to the Android operating system. Think about other ways to provide the `Context` to your implementation. 
+
+> :information_source: A `Context` is an interface to global information about an application environment. This is an abstract class whose implementation is provided by the Android system. It allows access to application-specific resources and classes, as well as up-calls for application-level operations such as launching activities, broadcasting and receiving intents, etc. (Source: [Android Documentation](https://developer.android.com/reference/android/content/Context))
+
+### HTTP requests
+
+There are a few options to access the network in Android. We suggest you to use the native `HttpsUrlConnection`. See the [Android Documentation](https://developer.android.com/training/basics/network-ops/connecting#httpsurlconnection) for a good example of how it works.
+
+But first, you need to disable a check. By default, Android disallows doing networking on main thread - as we explained, doing that is very bad. But since we will introduce asynchronicity later, you must add this code somewhere (for example in your main activity). This needs to run before the first call to your service.
+
+```java
+// TODO: Remove this line of code once I learned about asynchronous operations!
+StrictMode.setThreadPolicy(new StrictMode.ThreadPolicy.Builder().permitNetwork().build()); 
+```
+
+> Hint: you will need to add the `INTERNET` permission to access the net. To do so, add the following to your App Manifest (`AndroidManifest.xml`)
+> `<uses-permission android:name="android.permission.INTERNET" />`
+> To apply the change, **uninstall the app from the device/virtual device**. Otherwise, the permission won't be granted.
+
+> For advanced students: if you already know about asynchronicity, the easy way to do async networking in Android is to use [Volley](https://developer.android.com/training/volley/simple). We will probably introduce this tool later this semester.
+
+>Hint: we provide here a small example of HTTP request done in this fashion:
```

Comment:
Is this really necessary? Students are expected to spend 8-ish hours total on this step of the project (1.5 weeks at 4 credits minus the lecture time)

Reply:
No it is not, I will remove it